File: lib/src/calendars.js
/**
 * Calendars — per-year file loader.
 * Loads year JSON files from a directory path or base URL.
 */

import { BellSchedule } from './bell-schedule.js';

class Calendars {
  #basePath;
  #cache;

  /**
   * @param {string} basePath - Directory path (e.g. './calendars/') or URL base.
   */
  constructor(basePath) {
    this.#basePath = basePath;
    this.#cache = new Map(); // year string → parsed data
  }

  /**
   * Load data for a specific academic year.
   * @param {string} year - e.g. '2025-2026'
   * @returns {Promise<object[]>}
   */
  async #load(year) {
    if (this.#cache.has(year)) {
      return this.#cache.get(year);
    }

    const filePath = `${this.#basePath}${year}.json`;
    let data;

    if (this.#basePath.startsWith('http://') || this.#basePath.startsWith('https://')) {
      const res = await fetch(filePath);
      if (!res.ok) throw new Error(`Failed to fetch ${filePath}: ${res.status} ${res.statusText}`);
      data = await res.json();
    } else {
      const { readFile } = await import('node:fs/promises');
      const text = await readFile(filePath, 'utf8');
      data = JSON.parse(text);
    }

    // Normalize to array.
    const arr = Array.isArray(data) ? data : [data];
    this.#cache.set(year, arr);
    return arr;
  }

  /**
   * Get a BellSchedule for a specific academic year.
   * @param {string} year - e.g. '2025-2026'
   * @param {object} [options]
   * @returns {Promise<BellSchedule>}
   */
  async forYear(year, options = {}) {
    const arr = await this.#load(year);
    return new BellSchedule(arr, options);
  }

  /**
   * Get a BellSchedule appropriate for the current instant.
   * During summer, loads both the most recent ended year and the next upcoming
   * year so summer-bounds and next-year-start queries work correctly.
   * @param {object} [options]
   * @returns {Promise<BellSchedule>}
   */
  async current(options = {}) {
    const today = Temporal.Now.plainDateISO();
    const year = this.#academicYearFor(today);

    const primaryArr = await this.#load(year);
    const firstDay = primaryArr[0].firstDayTeachers || primaryArr[0].firstDay;
    const lastDay = primaryArr[0].lastDay;
    const inYear = today.toString() >= firstDay && today.toString() <= lastDay;

    if (inYear) {
      return new BellSchedule(primaryArr, options);
    }

    // Summer — load adjacent year.
    const allData = [...primaryArr];

    if (today.toString() > lastDay) {
      // After this year's end — load the next academic year.
      const nextYearLabel = this.#nextAcademicYear(year);
      try {
        const nextArr = await this.#load(nextYearLabel);
        allData.push(...nextArr);
      } catch {
        // Next year data not available; that's fine.
      }
    } else {
      // Before this year's start — load the previous academic year.
      const prevYearLabel = this.#prevAcademicYear(year);
      try {
        const prevArr = await this.#load(prevYearLabel);
        allData.unshift(...prevArr);
      } catch {
        // Previous year data not available; that's fine.
      }
    }

    return new BellSchedule(allData, options);
  }

  /**
   * Determine the academic year label for a given date.
   * Academic year starts in August.
   * @param {Temporal.PlainDate} date
   * @returns {string} e.g. '2025-2026'
   */
  #academicYearFor(date) {
    const { month, year } = date;
    if (month >= 8) {
      return `${year}-${year + 1}`;
    } else {
      return `${year - 1}-${year}`;
    }
  }

  /**
   * @param {string} year - e.g. '2025-2026'
   * @returns {string} - e.g. '2026-2027'
   */
  #nextAcademicYear(year) {
    const [start] = year.split('-').map(Number);
    return `${start + 1}-${start + 2}`;
  }

  /**
   * @param {string} year - e.g. '2025-2026'
   * @returns {string} - e.g. '2024-2025'
   */
  #prevAcademicYear(year) {
    const [start] = year.split('-').map(Number);
    return `${start - 1}-${start}`;
  }
}

export { Calendars };

File: libs/java/src/main/java/com/gigamonkeys/bells/Calendars.java
package com.gigamonkeys.bells;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.LocalDate;
import java.time.ZoneId;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Per-year calendar file loader. Loads {@code {year}.json} files (e.g.
 * {@code 2025-2026.json}) from a directory path or a base URL and builds {@link BellSchedule}
 * instances from them.
 */
public final class Calendars {

  private final String basePath;
  private final Map<String, List<CalendarData>> cache = new HashMap<>();
  private final HttpClient httpClient = HttpClient.newHttpClient();

  /**
   * @param basePath a directory path (e.g. {@code "./calendars/"}) or URL base
   *     (e.g. {@code "https://example.com/calendars/"})
   */
  public Calendars(String basePath) {
    this.basePath = basePath;
  }

  private List<CalendarData> load(String year) {
    List<CalendarData> cached = cache.get(year);
    if (cached != null) {
      return cached;
    }

    String location = basePath + year + ".json";
    String text;

    if (basePath.startsWith("http://") || basePath.startsWith("https://")) {
      try {
        HttpRequest request = HttpRequest.newBuilder(URI.create(location)).GET().build();
        HttpResponse<String> response =
            httpClient.send(request, HttpResponse.BodyHandlers.ofString());
        if (response.statusCode() < 200 || response.statusCode() >= 300) {
          throw new IOException("Failed to fetch " + location + ": HTTP " + response.statusCode());
        }
        text = response.body();
      } catch (IOException e) {
        throw new UncheckedIOException(e);
      } catch (InterruptedException e) {
        Thread.currentThread().interrupt();
        throw new RuntimeException("Interrupted while fetching " + location, e);
      }
    } else {
      try {
        text = Files.readString(Path.of(location), StandardCharsets.UTF_8);
      } catch (IOException e) {
        throw new UncheckedIOException(e);
      }
    }

    List<CalendarData> data = CalendarData.parse(text);
    cache.put(year, data);
    return data;
  }

  /**
   * Build a {@link BellSchedule} for a specific academic year, with default options.
   *
   * @param year e.g. {@code "2025-2026"}
   * @return the bell schedule
   */
  public BellSchedule forYear(String year) {
    return forYear(year, Options.defaults());
  }

  /**
   * Build a {@link BellSchedule} for a specific academic year.
   *
   * @param year e.g. {@code "2025-2026"}
   * @param options the viewer options
   * @return the bell schedule
   */
  public BellSchedule forYear(String year, Options options) {
    return new BellSchedule(load(year), options);
  }

  /**
   * Build a {@link BellSchedule} appropriate for the current instant, using the system-default
   * timezone to evaluate "today".
   *
   * @param options the viewer options
   * @return the bell schedule
   */
  public BellSchedule current(Options options) {
    return current(ZoneId.systemDefault(), options);
  }

  /**
   * Build a {@link BellSchedule} appropriate for the current instant. During summer, also
   * loads the adjacent year so summer-bounds and next-year-start queries work correctly.
   *
   * @param zone the timezone in which to evaluate "today"
   * @param options the viewer options
   * @return the bell schedule
   */
  public BellSchedule current(ZoneId zone, Options options) {
    LocalDate today = LocalDate.now(zone);
    String year = academicYearFor(today);

    List<CalendarData> primary = load(year);
    CalendarData first = primary.get(0);
    // Treat an empty firstDayTeachers as absent (matches the JS `|| firstDay`).
    String firstDay =
        (first.firstDayTeachers() == null || first.firstDayTeachers().isEmpty())
            ? first.firstDay()
            : first.firstDayTeachers();
    String lastDay = first.lastDay();
    String todayStr = today.toString();
    boolean inYear = todayStr.compareTo(firstDay) >= 0 && todayStr.compareTo(lastDay) <= 0;

    if (inYear) {
      return new BellSchedule(primary, options);
    }

    // Summer — load the adjacent year.
    List<CalendarData> allData = new java.util.ArrayList<>(primary);

    if (todayStr.compareTo(lastDay) > 0) {
      // After this year's end — load the next academic year.
      try {
        allData.addAll(load(nextAcademicYear(year)));
      } catch (RuntimeException ignored) {
        // Next year data not available; that's fine.
      }
    } else {
      // Before this year's start — load the previous academic year.
      try {
        allData.addAll(0, load(prevAcademicYear(year)));
      } catch (RuntimeException ignored) {
        // Previous year data not available; that's fine.
      }
    }

    return new BellSchedule(allData, options);
  }

  /**
   * Determine the academic-year label for a date. The academic year starts in August.
   *
   * @param date a date
   * @return e.g. {@code "2025-2026"}
   */
  static String academicYearFor(LocalDate date) {
    int month = date.getMonthValue();
    int year = date.getYear();
    if (month >= 8) {
      return year + "-" + (year + 1);
    } else {
      return (year - 1) + "-" + year;
    }
  }

  static String nextAcademicYear(String year) {
    int start = Integer.parseInt(year.split("-")[0]);
    return (start + 1) + "-" + (start + 2);
  }

  static String prevAcademicYear(String year) {
    int start = Integer.parseInt(year.split("-")[0]);
    return (start - 1) + "-" + start;
  }
}
